### ISSUE TITLE
getPublicNetworkNamePrefix return UNSAFE public network name

### ISSUE THREAD

**Issue by echoidcf** [apache/cloudstack#4247]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VR
~~~

##### CLOUDSTACK VERSION

~~~
All version is affected
~~~

##### CONFIGURATION

multiple public network like
vlan 11, vlan 110, vlan 1100
vmware

##### SUMMARY

in this file
cloudstack/plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java

<pre>
                String publicNeworkName = HypervisorHostHelper.getPublicNetworkNamePrefix(vlanId);
                Pair<Integer, VirtualDevice> publicNicInfo = vmMo.getNicDeviceIndex(publicNeworkName);
</pre>
You generate a public network name with a vlanId, and use this name to locate the nic device index.
The problem is getPublicNetworkNamePrefix works like this
<pre>
    public static String getPublicNetworkNamePrefix(String vlanId) {
        if (UNTAGGED_VLAN_NAME.equalsIgnoreCase(vlanId)) {
            return "cloud.public.untagged";
        } else {
            return "cloud.public." + vlanId;
        }
    }
</pre>
And will return something like "cloud.public.11" if vlan id is 11
getNicDeviceIndex only match the public network name with startWith which means this will match a network which vlan id is 111 or 110 or something like that.

You should return something like "cloud.public.11." to be safe.

##### STEPS TO REPRODUCE

~~~
create multiple public network with vlan 11,110,1101 etc.
and associate ip to vr.
~~~

##### EXPECTED RESULTS

~~~
should be normal
~~~

##### ACTUAL RESULTS

~~~
HIGH chance to associate ip to WRONG nic
~~~

**Comment by DaanHoogland** [comment]
sounds like you have figured the code @echoidcf . will you submit a PR for this?

**Comment by echoidcf** [comment]
> PR
What is a PR? Sorry, I am just newbie.

**Comment by rhtyd** [comment]
@echoidcf a pull request, sending code contribution to this repository

**Comment by rhtyd** [comment]
Ping @Spaceman1984 @DaanHoogland or @Pearl1594 @shwstppr - can you have a look at this, we already have example code?

**Comment by Pearl1594** [comment]
@rhtyd I can look into it

**Comment by rhtyd** [comment]
@Pearl1594 instead of the fix you've proposed can you fix the usage of `getNicDeviceIndex` to find exact or most matching portgroup instead of startswith(), i.e. the search should be exact instead of wildcard based which may pick a portgroup ending with 111 (VLAN ID 111) instead of 11 (VLAN ID 11).

**Comment by echoidcf** [comment]
> @echoidcf a pull request, sending code contribution to this repository
Sounds cool, but I don't know much about java. I can read, but coding is another thing ^^.

**Comment by rhtyd** [comment]
PR merged https://github.com/apache/cloudstack/pull/4254 - the fix will be in 4.15
